### PR TITLE
Enable MSVC W4 with a few exceptions. Fix warnings and bugs

### PR DESCRIPTION
**Pull request by yuslepukhin**

### CODE REVIEW

File: db/db_compaction_filter_test.cc [hunk 1]
```diff
@@ -308,7 +308,7 @@ TEST_F(DBTestCompactionFilter, CompactionFilter) {
     ASSERT_OK(iter->status());
     while (iter->Valid()) {
       ParsedInternalKey ikey(Slice(), 0, kTypeValue);
-      ikey.sequence = -1;
+      ikey.sequence = SequenceNumber(-1);
```
Comment (yiwu-arbug) [inline db/db_compaction_filter_test.cc]: SequenceNumber is alias of uint64_t. Seems to me we can simply remove this line.

File: db/db_compaction_filter_test.cc [hunk 2]
```diff
@@ -617,7 +617,7 @@ TEST_F(DBTestCompactionFilter, CompactionFilterContextManual) {
...
```
Comment (yiwu-arbug) [inline db/db_compaction_filter_test.cc]: ditto

File: db/db_test_util.cc [hunk 3]
```diff
@@ -1169,7 +1169,7 @@ void DBTestBase::validateNumberOfEntries(int numValues, int cf) {
   int seq = numValues;
   while (iter->Valid()) {
     ParsedInternalKey ikey;
```
Comment (yiwu-arbug) [inline db/db_test_util.cc]: initialize ikey with sequence = 0 with constructor, then remove the next line.

File: tools/ldb_cmd.cc [hunk 4]
```diff
@@ -2806,7 +2806,7 @@ void DumpSstFile(std::string filename, bool output_hex, bool show_properties) {
   }
   // no verification
   rocksdb::SstFileReader reader(filename, false, output_hex);
-  Status st = reader.ReadSequential(true, -1, false,  // has_from
+  Status st = reader.ReadSequential(true, static_cast<uint64_t>(-1), false,  // has_from
```
Comment (yiwu-arbug) [inline tools/ldb_cmd.cc]: should be `std::numeric_limits<uint64_t>::max()`.

File: tools/sst_dump_tool.cc [hunk 5]
```diff
@@ -411,7 +411,7 @@ void print_help() {
 
 int SSTDumpTool::Run(int argc, char** argv) {
   const char* dir_or_file = nullptr;
-  uint64_t read_num = -1;
+  uint64_t read_num = uint64_t(-1);
```
Comment (yiwu-arbug) [inline tools/sst_dump_tool.cc]: ditto

File: util/coding.cc [hunk 6]
```diff
@@ -15,6 +15,7 @@
 
 namespace rocksdb {
 
+#pragma warning(disable : 4244)
```
Comment (yiwu-arbug) [inline util/coding.cc]: comment on what warning is disabled.

File: utilities/blob_db/blob_log_writer.cc [hunk 7]
```diff
@@ -88,7 +88,7 @@ Status Writer::AddRecord(const Slice& key, const Slice& val,
   assert(last_elem_type_ == kEtFileHdr || last_elem_type_ == kEtFooter);
 
   std::string buf;
-  ConstructBlobHeader(&buf, key, val, -1, -1);
+  ConstructBlobHeader(&buf, key, val, uint64_t(-1), -1);
```
Comment (yiwu-arbug) [inline utilities/blob_db/blob_log_writer.cc]: Just put 0 here.

File: utilities/redis/redis_lists_test.cc [hunk 8]
```diff
@@ -747,7 +747,7 @@ namespace {
 void MakeUpper(std::string* const s) {
   int len = static_cast<int>(s->length());
   for (int i = 0; i < len; ++i) {
-    (*s)[i] = toupper((*s)[i]);  // C-version defined in <ctype.h>
+    (*s)[i] = char(toupper((*s)[i]));  // C-version defined in <ctype.h>
```
Comment (yiwu-arbug) [inline utilities/redis/redis_lists_test.cc]: `static_cast<char>`

File: utilities/spatialdb/spatial_db_test.cc [hunk 9]
```diff
@@ -224,7 +224,7 @@ namespace {
 std::string RandomStr(Random* rnd) {
   std::string r;
   for (int k = 0; k < 10; ++k) {
-    r.push_back(rnd->Uniform(26) + 'a');
+    r.push_back(char(rnd->Uniform(26)) + 'a');
```
Comment (yiwu-arbug) [inline utilities/spatialdb/spatial_db_test.cc]: `static_cast<char>`

File: utilities/transactions/transaction_test.cc [hunk 10]
```diff
@@ -1153,10 +1153,10 @@ TEST_P(TransactionTest, DISABLED_TwoPhaseMultiThreadTest) {
     if (id % 2 == 0) {
       txn_options.expiration = 1000000;
     }
-    TransactionName name("xid_" + std::string(1, 'A' + id));
+    TransactionName name("xid_" + std::string(1, 'A' + char(id)));
```
Comment (yiwu-arbug) [inline utilities/transactions/transaction_test.cc]: `static_cast<char>`

File: utilities/transactions/transaction_test.cc [hunk 11]
```diff
@@ -1208,8 +1208,8 @@ TEST_P(TransactionTest, DISABLED_TwoPhaseMultiThreadTest) {
   std::string value;
   Status s;
   for (uint32_t t = 0; t < NUM_TXN_THREADS; t++) {
-    TransactionName name("xid_" + std::string(1, 'A' + t));
-    for (int i = 0; i < 10; i++) {
+    TransactionName name("xid_" + std::string(1, 'A' + char(t)));
```
Comment (yiwu-arbug) [inline utilities/transactions/transaction_test.cc]: ditto

File: db/db_test_util.cc (hunk 3 above)
Comment (yuslepukhin) [reply db/db_test_util.cc]: Would .clear() be fine?

File: third-party/fbson/FbsonWriter.h [hunk 12]
```diff
@@ -32,6 +32,8 @@
 #include "FbsonDocument.h"
 #include "FbsonStream.h"
 
+#pragma warning(disable : 4244)
```
Comment (Orvid) [inline third-party/fbson/FbsonWriter.h]: This should probably be re-enabled at the end of the header, shouldn't it?

File: db/db_test_util.cc (hunk 3 above)
Comment (yiwu-arbug) [reply db/db_test_util.cc]: Maybe just remove the next line. Let see if clang_analyzer complains..

File: third-party/fbson/FbsonWriter.h (hunk 12 above)
Comment (yuslepukhin) [reply third-party/fbson/FbsonWriter.h]: This is a template, so the warnings are emitted at the point of instantiation. so can not do that. A better idea would be great. At this point I decided not a big deal"

`third-party\fbson\FbsonJsonParser.h:#include "FbsonWriter.h"
third-party\fbson\FbsonWriter.h:#ifndef FBSON_FBSONWRITER_H
third-party\fbson\FbsonWriter.h:#define FBSON_FBSONWRITER_H
third-party\fbson\FbsonWriter.h:#endif // FBSON_FBSONWRITER_H
utilities\document\json_document_builder.cc:#include "third-party/fbson/FbsonWriter.h"`

**Comment by yiwu-arbug** [comment]
Getting compile error on linux: https://gist.github.com/yiwu-arbug/8bc8cf08dbb0eca45f85139c74d32729

**Comment by yiwu-arbug** [comment]
Also, mind running `make format` once to fix formatting?

**Comment by yuslepukhin** [comment]
@yiwu-arbug With regards to the error 1) This is non-standard. Compilers are supposed to ignore unfamiliar `pragmas`. Is it possible to disable that warning?
2) I do not have Linux build env.

**Comment by yiwu-arbug** [comment]
@yuslepukhin can you try update Makefile, CMake and in particular, the ./TARGETS file with -Wno-unknown-pragmas and update the PR?

**Comment by yiwu-arbug** [comment]
@yuslepukhin now it hits the following error: https://gist.github.com/yiwu-arbug/49941767eeec8fdcdad35c740882f108

**Comment by yiwu-arbug** [comment]
Travis cmake-mingw is failing.

**Pull request closed without merging**